e2e/tests/stablex_test.rs

```diff
@@ -111,3 +120,124 @@ fn test_stablex() {
+    eloop.into_remote();
+    let web3 = Web3::new(http);
+    let mut instance = BatchExchange::deployed(&web3)
+        .wait()
+        .expect("Cannot get deployed Batch Exchange");
+    let secret = {
+        let private_key: H256 = env::var("PK")
+            .expect("PK env var not set")
+            .parse()
+            .expect("PK not parseable");
+        SecretKey::from_raw(&private_key[..])
+            .map_err(ethsign::Error::from)
+            .expect("Cannot derive key")
+    };
+    let account = Account::Offline(secret, None);
+    instance.defaults_mut().from = Some(account.clone());
+    instance.defaults_mut().gas = Some(1_000_000.into());
+    instance.defaults_mut().gas_price = Some(8_000_000_000u64.into());
+
+    let nonce = web3
+        .eth()
+        .transaction_count(account.address(), None)
+        .wait()
+        .expect("Cannot get nonce");
+    println!("Using account {:x} with nonce {}", account.address(), nonce);
+
+    // Gather token and batch info
+    let token_a = instance
+        .token_id_to_address_map(0)
+        .call()
+        .wait()
+        .expect("Cannot get first Token address");
+    let token_b = instance
+        .token_id_to_address_map(7)
+        .call()
+        .wait()
+        .expect("Cannot get second Token address");
+    let batch = instance
+        .get_current_batch_id()
+        .call()
+        .wait()
+        .expect("Cannot get batchId");
+
+    // Approve Funds
+    let first_approve = IERC20::at(&web3, token_a)
+        .approve(instance.address(), 1_000_000.into())
+        .nonce(nonce)
+        .gas(1_000_000.into())
+        .gas_price(8_000_000_000u64.into())
+        .from(account.clone())
+        .send_and_confirm(Duration::from_secs(1), 1);
+    let second_approve = IERC20::at(&web3, token_b)
+        .approve(instance.address(), 1_000_000.into())
+        .nonce(nonce + 1)
+        .gas(1_000_000.into())
+        .gas_price(8_000_000_000u64.into())
+        .from(account.clone())
+        .send_and_confirm(Duration::from_secs(1), 1);
+
+    // Deposit Funds
+    let first_deposit = instance
+        .deposit(token_a, 1_000_000.into())
+        .nonce(nonce + 2)
+        .send_and_confirm(Duration::from_secs(1), 1);
+    let second_deposit = instance
+        .deposit(token_b, 1_000_000.into())
+        .nonce(nonce + 3)
+        .send_and_confirm(Duration::from_secs(1), 1);
+
+    // Place orders
+    let first_order = instance
+        .place_order(0, 7, batch + 2, 1_000_000.into(), 10_000_000.into())
+        .nonce(nonce + 4)
+        .send_and_confirm(Duration::from_secs(1), 1);
+    let second_order = instance
+        .place_order(7, 0, batch + 1, 1_000_000.into(), 10_000_000.into())
+        .nonce(nonce + 5)
+        .send_and_confirm(Duration::from_secs(1), 1);
+
+    // Wait for all transactions to be confirmed
+    println!("Waiting for transactions to be confirmed");
+    let results = join_all(vec![
```

Comment:
I like `try_join!` for this, it also turns it into a `Result<(T1, T2, T3, ...), Error>` for you.

Reply:
The problem with try_join is that it itself awaits the futures and thus can only be used in an async function. There is try_join_all which does the same as try_join but returns a Future instead of directly awaiting it.

However, I'd prefer knowing which tx failed and thus unwrapping per item helps with that. I agree, that once we log the function name in the Future Error this may no longer be needed.